userland/userland_generic.ld

```diff
@@ -24,36 +24,57 @@ SECTIONS {
     .text :
     {
         _text = .;
+        /**
+         * Populate the header expected by `crt0`:
+         *
+         *
+         *  struct hdr {
+         *    uint32_t got_sym_start;
+         *    uint32_t got_start;
+         *    int got_size;
+         *    uint32_t data_sym_start;
+         *    uint32_t data_start;
+         *    int data_size;
+         *    uint32_t bss_start;
+         *    int bss_size;
+         *    uint32_t reldata_start;
+         *  };
+         */
+        /* Offset of GOT symbols in flash */
+        LONG(LOADADDR(.got) - _text);
+        /* Offset of GOT section in memory */
+        LONG(_got);
+        /* Size of GOT section */
+        LONG(SIZEOF(.got));
+        /* Offset of data symbols in flash */
+        LONG(LOADADDR(.data) - _text);
+        /* Offset of data section in memory */
+        LONG(_data);
+        /* Size of data section */
+        LONG(SIZEOF(.data));
+        /* Offset of BSS section in memory */
+        LONG(_sbss);
+        /* Size of BSS section */
+        LONG(SIZEOF(.bss));
+        /* First address offset after program flash, where elf2tbf places
+         * .rel.data section */
+        LONG(LOADADDR(.endsec) - _text);
```

Comment:
Probably I am not understanding the linker script correctly but why do we include a bunch of `LONG`s while the `hdr` struct consists of a mixture of `uint32_t` and `int`s. Is it because of the sign? And isn't `int` architecture dependent or do we redefine it somehow?

Reply:
`LONG` is a word-sized integer in ARM linker script speak. Mixing between `uint32_t` and `int` indeed doesn't make a ton of sense. It was to make the compiler happier about comparing to `ints` in the for-loops, but maybe better to just make the iterator variable a `uint32_t`. What do you think @Woyten?

Reply:
Personally, I prefer the version that generates the least amount of questions to the reader of the code. Therefore, I would definitely vote for making everything a `uint32_t`. However, the compiler warnings are a bad thing, too. There should be a way to suppress them if this is still necessary.

Reply:
Oh, I was unclear... the compiler warnings are gone if I just make the for-loop iterator (`i`) and `uint32_t` instead of an `int`, which is also perfectly fine. See the commit I just pushed.

Reply:
Nice! :smiley: